### PR TITLE
Framework: Extract "edit-post" to its own module

### PR DESCRIPTION
**Pull request by youknowriad**
This PR extracts the `edit-post` module to its own module. This step is helpful to be able to experiment with an "edit-template" later.

 - The new module has its own store handling layout state.
 - Extract some more reusable components in `editor` for use in `edit-post` (`BlockSelectionClearer`, `InserterWithShortcuts`...)
 - The "Show block inspector" button proved to be a challenge, because it's layout related. this is fixed by providing a way to extend the Block's menu from outside the `editor` module using a prop.
 - Renamed classNames for the components moved to the new module.

**Testing instructions**

 - This may introduce some unnoticed regressions, heavily testing this PR is necessary.

### CODE REVIEW

File: lib/client-assets.php [hunk 1]
```diff
@@ -159,13 +159,19 @@ function gutenberg_register_scripts_and_styles() {
...
 		array( 'postbox', 'jquery', 'wp-api', 'wp-data', 'wp-date', 'wp-i18n', 'wp-blocks', 'wp-element', 'wp-components', 'wp-utils', 'word-count', 'editor', 'heartbeat' ),
-		filemtime( gutenberg_dir_path() . 'editor/build/index.js' ),
-		true // enqueue in the footer.
+		filemtime( gutenberg_dir_path() . 'editor/build/index.js' )
+	);
+
+	wp_register_script(
+		'wp-edit-post',
+		gutenberg_url( 'edit-post/build/index.js' ),
+		array( 'wp-element', 'wp-components', 'wp-editor', 'wp-i18n', 'wp-date', 'wp-utils' ),
```
Comment (aduth) [inline lib/client-assets.php]: `edit-post/index.js` has a reference to jQuery and heartbeat, so this should be made an explicit dependency.

Heartbeat is no longer used as a dependency in `wp-editor`.

File: webpack.config.js [hunk 2]
```diff
@@ -53,6 +53,7 @@ const entryPointNames = [
 	'i18n',
 	'utils',
 	'data',
+	'edit-post',
```
Comment (aduth) [inline webpack.config.js]: To be able to export this as `wp.editPost` we should consider mapping packages using [Lodash's `_.camelCase`](https://lodash.com/docs/4.17.4#camelCase).

File: package.json [hunk 3]
```diff
@@ -104,15 +104,15 @@
 	},
 	"jest": {
 		"collectCoverageFrom": [
-			"(blocks|components|date|editor|element|i18n|data|utils)/**/*.js"
```
Comment (aduth) [inline package.json]: We also include a custom ESLint rule for each top-level directory in `.eslintrc.json`:

https://github.com/WordPress/gutenberg/blob/e694f6a99b7ce01bb3586798e1af6c662a7a4895/.eslintrc.json#L95-L125

**Review by aduth — COMMENTED** [review]
I agree with general direction that this helps us toward consolidating `editor` and `blocks` directory as also noted in #2761. There's a few open questions remaining on state responsible for editor vs. `edit-post` (e.g. `currentPost` remains in editor state) that we could consider refactoring further in subsequent pull requests.

The addition of `storeKey` to each connected component is quite verbose. I think we should consider either a `connect` custom to each top-level module which serves as a wrapper to React-Redux's `connect` passing the `storeKey` automatically. Alternatively, we could seek to enhance our own `query` HoC to support all of the same functionality as `connect` and then it would be not necessary.

Generally difficult to review some of the more granular changes, as while some files are moved verbatim, others include revisions.

File: webpack.config.js (hunk 2 above)
Comment (youknowriad) [reply webpack.config.js]: Might not be possible to configure webpack to do so https://webpack.js.org/configuration/output/#output-library

**Comment by youknowriad** [comment]
> Alternatively, we could seek to enhance our own query HoC to support all of the same functionality as connect and then it would be not necessary.

If we're fine exposing all selectors to third-party modules, we can get rid of the `connect` usage.

--

I addressed most of the feedback. Thanks @aduth

File: editor/components/block-list/block.js [hunk 4]
```diff
@@ -433,7 +433,7 @@ export class BlockListBlock extends Component {
 			>
 				<BlockDropZone index={ order } />
 				{ ( showUI || isHovered ) && <BlockMover uids={ [ block.uid ] } /> }
-				{ ( showUI || isHovered ) && <BlockSettingsMenu uids={ [ block.uid ] } /> }
+				{ ( showUI || isHovered ) && <BlockSettingsMenu uids={ [ block.uid ] } renderBlockMenu={ renderBlockMenu } /> }
```
Comment (aduth) [inline editor/components/block-list/block.js]: This change with `renderBlockMenu` broke the multi-select controls dropdown:

1. Create a multi-selection range
2. Open "More Options" menu

multi-select missing

Comment (youknowriad) [reply editor/components/block-list/block.js]: Good catch, we should cherry-pick the fix from the templates branch to its own PR

**Comment by aduth** [comment]
Nitpick: To the naming of this module, it stands out to me as inconsistent with our general naming, not only because it's intended as a superset of functionality for the base `editor` (but naming does not imply this), but also that other modules generally tend to be nouns (`editor`, `blocks`, `components`), whereas `edit-post` is more an action/verb.

My suggested alternative would be `post-editor`. What do you think?

**Comment by aduth** [comment]
Also, we should plan to fix the assignment of the global to avoid `wp[ 'edit-post' ]` and instead as `wp.editPost`. I agree it wasn't a blocker for this pull request, but I don't think we should settle on the non-camelCased global property.

**Comment by youknowriad** [comment]
cc @mtias for the name of the module.

About the global, this might mean we need to find another way to assign these variables (instead of relying on the library config of webpack)

File: editor/components/block-list/index.js [hunk 5]
```diff
@@ -245,10 +246,10 @@ class BlockList extends Component {
 	}
 
 	render() {
-		const { blocks, showContextualToolbar } = this.props;
+		const { blocks, showContextualToolbar, renderBlockMenu } = this.props;
 
 		return (
-			<div>
+			<BlockSelectionClearer>
```
Comment (aduth) [inline editor/components/block-list/index.js]: Why was this changed from a `div` to `BlockSelectionClearer` ?

Comment (youknowriad) [reply editor/components/block-list/index.js]: Because this was handled in the parent component by using a ref to this div there. I just made this component generic and used it in both the `VisualEditor` component and the inner component `BlockList`

**Comment by jasmussen** [comment]
I think this PR regressed the color scheme styling of the tabs in the sidebar:

screen shot 2018-01-31 at 10 20 42

As you can see from the Publish button, it's supposed to be red, like other tab buttons:

screen shot 2018-01-31 at 10 22 17

It should be a fairly easy fix, renaming some CSS classes in edit-post/assets/stylesheets/_admin-schemes.scss.

Do you want to do it, @youknowriad, or would you like me to do it? I have time if you're busy.
